safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csp
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// Directive is the name of a single CSP directive.
+type Directive string
+
+const (
+	DirectiveScriptSrc Directive = "script-src"
+	DirectiveStyleSrc  Directive = "style-src"
+	DirectiveObjectSrc Directive = "object-src"
+	DirectiveBaseURI   Directive = "base-uri"
+	DirectiveReportURI Directive = "report-uri"
+)
+
+const (
+	ValueHTTPS         = "https:"
+	ValueHTTP          = "http:"
+	ValueUnsafeEval    = "'unsafe-eval'"
+	ValueUnsafeInline  = "'unsafe-inline'"
+	ValueNone          = "'none'"
+	ValueStrictDynamic = "'strict-dynamic'"
+)
+
+// PolicyDirective contains a single CSP directive.
+type PolicyDirective struct {
+	dir       Directive
+	vals      []string
+	nonces    bool
+	lastNonce string
+	// readRand is used for dependency injection in tests.
+	readRand func([]byte) (int, error)
+}
+
+// NewPolicyDirective creates a new PolicyDirective given the directive, the
+// values and whether or not a nonce directive should be added.
+func NewPolicyDirective(d Directive, values []string, nonces bool) *PolicyDirective {
+	return &PolicyDirective{
+		dir:      d,
+		vals:     values,
+		nonces:   nonces,
+		readRand: rand.Read,
+	}
+}
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 8
+
+func (pd *PolicyDirective) generateNonce() {
+	b := make([]byte, nonceSize)
+	_, err := pd.readRand(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	pd.lastNonce = base64.StdEncoding.EncodeToString(b)
+}
```

Comment:
This doesn't use `PolicyDirective` at all expect to save `lastNonce`, which probably shouldn't even belong to the directive type. Make it a free function that accepts the rand.Read as an argument.

Make it return the nonce, instead of saving it.